sfe/sfe.go

```diff
@@ -0,0 +1,317 @@
+	rapb "github.com/letsencrypt/boulder/ra/proto"
+	sapb "github.com/letsencrypt/boulder/sa/proto"
+)
+
+const (
+	// The API version should be checked when parsing parameters to quickly deny
+	// a client request. Can be used to mass-invalidate URLs. Must be
+	// concatenated with other path slugs.
+	unpauseAPISlug = "/sfe/v1"
+)
+
+var (
+	unpauseGetForm    = unpauseAPISlug + "/unpause"
+	unpausePostForm   = unpauseAPISlug + "/do-unpause"
+	unpauseSuccessful = unpauseAPISlug + "/unpause-successful"
+
+	//go:embed all:static
+	staticFS embed.FS
+
+	//go:embed all:templates all:pages
+	dynamicFS embed.FS
+)
+
+// SelfServiceFrontEndImpl provides all the logic for Boulder's selfservice
+// frontend web-facing interface, i.e., a portal where a subscriber can unpause
+// their account. Its methods are primarily handlers for HTTPS requests for the
+// various non-ACME functions.
+type SelfServiceFrontEndImpl struct {
+	ra rapb.RegistrationAuthorityClient
+	sa sapb.StorageAuthorityReadOnlyClient
+
+	log blog.Logger
+	clk clock.Clock
+
+	// requestTimeout is the per-request overall timeout.
+	requestTimeout time.Duration
+
+	// unpauseHMACHash is used to validate incoming JWT signatures on the
+	// unpause endpoint and should be shared by the SFE and WFE.
+	unpauseHMACHash []byte
+
+	// HTML pages served by the SFE
+	templatePages *template.Template
+}
+
+// NewSelfServiceFrontEndImpl constructs a web service for Boulder
+func NewSelfServiceFrontEndImpl(
+	stats prometheus.Registerer,
+	clk clock.Clock,
+	logger blog.Logger,
+	requestTimeout time.Duration,
+	rac rapb.RegistrationAuthorityClient,
+	sac sapb.StorageAuthorityReadOnlyClient,
+	unpauseHMACHash []byte,
+) (SelfServiceFrontEndImpl, error) {
+
+	// Parse the files once at startup to avoid each request causing the server
+	// to JIT parse. The pages are stored in an in-memory embed.FS to prevent
+	// unnecessary filesystem I/O on a physical HDD.
+	tmplPages := template.Must(template.New("pages").ParseFS(dynamicFS, "templates/layout.html", "pages/*"))
+
+	sfe := SelfServiceFrontEndImpl{
+		log:             logger,
+		clk:             clk,
+		requestTimeout:  requestTimeout,
+		ra:              rac,
+		sa:              sac,
+		unpauseHMACHash: unpauseHMACHash,
+		templatePages:   tmplPages,
+	}
+
+	return sfe, nil
+}
+
+// Handler returns an http.Handler that uses various functions for various
+// non-ACME-specified paths. Each endpoint should have a corresponding HTML
+// page that shares the same name as the endpoint.
+func (sfe *SelfServiceFrontEndImpl) Handler(stats prometheus.Registerer, oTelHTTPOptions ...otelhttp.Option) http.Handler {
+	m := http.NewServeMux()
+
+	sfs, _ := fs.Sub(staticFS, "static")
+	staticAssetsHandler := http.StripPrefix("/static/", http.FileServerFS(sfs))
+
+	m.Handle("GET /static/", staticAssetsHandler)
+	m.HandleFunc("/", sfe.Index)
+	m.HandleFunc("GET /build", sfe.BuildID)
+	m.HandleFunc(unpauseGetForm, sfe.Unpause)
+	m.HandleFunc(unpausePostForm, sfe.Unpause)
```

Comment:
Now that we have different URLs for these endpoints, they should also call different functions. `getHelper` should be renamed something like `UnpauseForm` and specified in the first of these, and `postHelper` should be renamed something like `UnpauseSubmit` and specified in the second of these.